File: crates/nil-lua/src/client/cheat/infrastructure.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::client::ClientUserData;
use mlua::{LuaSerdeExt, UserDataMethods, Value};
use nil_payload::cheat::infrastructure::*;

pub fn add_methods<M: UserDataMethods<ClientUserData>>(methods: &mut M) {
  methods.add_async_method(
    "cheatGetAcademyRecruitQueue",
    async |lua, this, req: Value| {
      let req: CheatGetAcademyRecruitQueueRequest = lua.from_value(req)?;
      this
        .client(async |it| it.cheat_get_academy_recruit_queue(req).await)
        .await
        .map(|it| lua.to_value(&it))?
    },
  );

  methods.add_async_method(
    "cheatGetAcademyRecruitQueues",
    async |lua, this, req: Value| {
      let req: CheatGetAcademyRecruitQueuesRequest = lua.from_value(req)?;
      this
        .client(async |it| {
          it.cheat_get_academy_recruit_queues(req)
            .await
        })
        .await
        .map(|it| lua.to_value(&it))?
    },
  );

  methods.add_async_method(
    "cheatGetAllAcademyRecruitQueues",
    async |lua, this, req: Value| {
      let req: CheatGetAllAcademyRecruitQueuesRequest = lua.from_value(req)?;
      this
        .client(async |it| {
          it.cheat_get_all_academy_recruit_queues(req)
            .await
        })
        .await
        .map(|it| lua.to_value(&it))?
    },
  );

  methods.add_async_method(
    "cheatGetAllPrefectureBuildQueues",
    async |lua, this, req: Value| {
      let req: CheatGetAllPrefectureBuildQueuesRequest = lua.from_value(req)?;
      this
        .client(async |it| {
          it.cheat_get_all_prefecture_build_queues(req)
            .await
        })
        .await
        .map(|it| lua.to_value(&it))?
    },
  );

  methods.add_async_method(
    "cheatGetAllStableRecruitQueues",
    async |lua, this, req: Value| {
      let req: CheatGetAllStableRecruitQueuesRequest = lua.from_value(req)?;
      this
        .client(async |it| {
          it.cheat_get_all_stable_recruit_queues(req)
            .await
        })
        .await
        .map(|it| lua.to_value(&it))?
    },
  );

  methods.add_async_method("cheatGetInfrastructure", async |lua, this, req: Value| {
    let req: CheatGetInfrastructureRequest = lua.from_value(req)?;
    this
      .client(async |it| it.cheat_get_infrastructure(req).await)
      .await
      .map(|it| lua.to_value(&it))?
  });

  methods.add_async_method(
    "cheatGetPrefectureBuildQueue",
    async |lua, this, req: Value| {
      let req: CheatGetPrefectureBuildQueueRequest = lua.from_value(req)?;
      this
        .client(async |it| {
          it.cheat_get_prefecture_build_queue(req)
            .await
        })
        .await
        .map(|it| lua.to_value(&it))?
    },
  );

  methods.add_async_method(
    "cheatGetPrefectureBuildQueues",
    async |lua, this, req: Value| {
      let req: CheatGetPrefectureBuildQueuesRequest = lua.from_value(req)?;
      this
        .client(async |it| {
          it.cheat_get_prefecture_build_queues(req)
            .await
        })
        .await
        .map(|it| lua.to_value(&it))?
    },
  );

  methods.add_async_method(
    "cheatGetStableRecruitQueue",
    async |lua, this, req: Value| {
      let req: CheatGetStableRecruitQueueRequest = lua.from_value(req)?;
      this
        .client(async |it| it.cheat_get_stable_recruit_queue(req).await)
        .await
        .map(|it| lua.to_value(&it))?
    },
  );

  methods.add_async_method(
    "cheatGetStableRecruitQueues",
    async |lua, this, req: Value| {
      let req: CheatGetStableRecruitQueuesRequest = lua.from_value(req)?;
      this
        .client(async |it| it.cheat_get_stable_recruit_queues(req).await)
        .await
        .map(|it| lua.to_value(&it))?
    },
  );

  methods.add_async_method("cheatGetStorageCapacity", async |lua, this, req: Value| {
    let req: CheatGetStorageCapacityRequest = lua.from_value(req)?;
    this
      .client(async |it| it.cheat_get_storage_capacity(req).await)
      .await
      .map(|it| lua.to_value(&it))?
  });

  methods.add_async_method("cheatSetBuildingLevel", async |lua, this, req: Value| {
    let req: CheatSetBuildingLevelRequest = lua.from_value(req)?;
    this
      .client(async |it| it.cheat_set_building_level(req).await)
      .await
      .map(|it| lua.to_value(&it))?
  });

  methods.add_async_method(
    "cheatSetMaxInfrastructure",
    async |lua, this, req: Value| {
      let req: CheatSetMaxInfrastructureRequest = lua.from_value(req)?;
      this
        .client(async |it| it.cheat_set_max_infrastructure(req).await)
        .await
        .map(|it| lua.to_value(&it))?
    },
  );
}

File: app/src/types/request/cheat/infrastructure.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import type { Ruler } from '@/types/core/ruler';
import type { WorldId } from '@/types/core/world';
import type { Coord } from '@/types/core/continent';
import type { BuildingId, BuildingLevel } from '@/types/core/infrastructure/building';

export interface CheatGetAcademyRecruitQueueRequest {
  readonly world: WorldId;
  readonly coord: Coord;
}

export interface CheatGetAcademyRecruitQueuesRequest {
  readonly world: WorldId;
  readonly coords: readonly Coord[];
  readonly filterEmpty: boolean;
}

export interface CheatGetAllAcademyRecruitQueuesRequest {
  readonly world: WorldId;
  readonly filterEmpty: boolean;
}

export interface CheatGetAllPrefectureBuildQueuesRequest {
  readonly world: WorldId;
  readonly filterEmpty: boolean;
}

export interface CheatGetAllStableRecruitQueuesRequest {
  readonly world: WorldId;
  readonly filterEmpty: boolean;
}

export interface CheatGetInfrastructureRequest {
  readonly world: WorldId;
  readonly coord: Coord;
}

export interface CheatGetPrefectureBuildQueueRequest {
  readonly world: WorldId;
  readonly coord: Coord;
}

export interface CheatGetPrefectureBuildQueuesRequest {
  readonly world: WorldId;
  readonly coords: readonly Coord[];
  readonly filterEmpty: boolean;
}

export interface CheatGetStableRecruitQueueRequest {
  readonly world: WorldId;
  readonly coord: Coord;
}

export interface CheatGetStableRecruitQueuesRequest {
  readonly world: WorldId;
  readonly coords: readonly Coord[];
  readonly filterEmpty: boolean;
}

export interface CheatGetStorageCapacityRequest {
  readonly world: WorldId;
  readonly ruler: Option<Ruler>;
}

export interface CheatSetBuildingLevelRequest {
  readonly world: WorldId;
  readonly coord: Coord;
  readonly id: BuildingId;
  readonly level: BuildingLevel;
}

export interface CheatSetMaxInfrastructureRequest {
  readonly world: WorldId;
  readonly coord: Coord;
}
